src/pudl/helpers.py

```diff
@@ -332,90 +332,271 @@ def is_doi(doi):
+
+    suffixes = ["", ""]
+    if left_date_col == right_date_col:
+        if "suffixes" in kwargs:
+            suffixes = kwargs["suffixes"]
+        else:
+            suffixes = ["_x", "_y"]
+    # reconstruct the new report date column and clean up columns
+    left_right_date_col = [left_date_col + suffixes[0], right_date_col + suffixes[1]]
+    if report_at_start:
+        # keep the later of the two report dates when determining
+        # the new report date for each row
+        reconstructed_date = out[left_right_date_col].max(axis=1)
+    else:
+        # keep the earlier of the two report dates
+        reconstructed_date = out[left_right_date_col].min(axis=1)
+    out = out.drop(left_right_date_col + date_on, axis=1)
+    out.insert(loc=0, column=new_date_col, value=reconstructed_date)
+    return out
+
+
+def expand_timeseries(
+    df: pd.DataFrame,
+    key_cols: List[str],
+    date_col: str = "report_date",
+    freq: str = "MS",
+    fill_through_freq: Literal["year", "month", "day"] = "year",
+) -> pd.DataFrame:
+    """Expand a dataframe to a include a full time series at a given frequency.
+
+    This function adds a full timeseries to the given dataframe for each group
+    of columns specified by ``key_cols``. The data in the timeseries will be filled
+    with the next previous chronological observation for a group of primary key columns
+    specified by ``key_cols``.
+
+    Arguments:
+        df: The dataframe to expand. Must have ``date_col`` in columns.
+        key_cols: Column names of the non-date primary key columns in the dataframe.
+            The resulting dataframe will have a full timeseries expanded for each
+            unique group of these ID columns that are present in the dataframe.
+        date_col: Name of the datetime column being expanded into a full timeseries.
+        freq: The frequency of the time series to expand the data to.
+            See :ref:`here <timeseries.offset_aliases>` for a list of
+            frequency aliases.
+        fill_through_freq: The frequency in which to fill in the data through. For
+            example, if equal to "year" the data will be filled in through the end of
+            the last reported year for each grouping of `key_cols`. Valid frequencies
+            are only year, month, or day.
+    """
+    try:
+        pd.tseries.frequencies.to_offset(freq)
+    except ValueError:
+        logger.exception(
+            f"Frequency string {freq} is not valid. \
+            See Pandas Timeseries Offset Aliases docs for valid strings."
+        )
+
+    # For each group of ID columns add a dummy record with the date column
+    # equal to one increment higher than the last record in the group for the
+    # desired fill_through_freq.
+    # This allows records to be filled through the end of the last reported period
+    # and then this dummy record is dropped
+    df = convert_col_to_datetime(df, date_col)
```

Comment:
I thinkkk this `convert_col_to_datetime` could just happen on `end_dates` (after the `groupby` call one line later) and it would be faster. But this seemed safer? In case report date strings were formatted weirdly and `groupby` + `max` didn't get the latest date?